apps/emqx_durable_storage/src/emqx_ds_snapshot_manager.erl

```diff
@@ -0,0 +1,428 @@
+    },
+    case ra_log_snapshot:recover(Dir) of
+        {ok, Meta, MachineState} ->
+            ShardId = shard_id(MachineState),
+            RS1 = RS0#rs{state = MachineState},
+            ok = gen_server:call(?via(ShardId), #start_reader{rs = RS1, caller = self()}, infinity),
+            ReadState = #{shard_id => ShardId},
+            {ok, Meta, ReadState};
+        Error ->
+            Error
+    end.
+
+-spec read_chunk(ext_rs(), _Size :: non_neg_integer(), _SnapshotDir :: file:filename()) ->
+    {ok, binary(), {next, ext_rs()} | last} | {error, _Reason :: term()}.
+read_chunk(#{shard_id := ShardId} = ExtRS, Size, _Dir) ->
+    case gen_server:call(?via(ShardId), #read_chunk{size = Size}, infinity) of
+        {ok, Chunk, next} ->
+            ?tp(dsrepl_read_chunk_next, #{}),
+            {ok, Chunk, {next, ExtRS}};
+        {ok, Chunk, last} ->
+            {ok, Chunk, last};
+        {error, Reason} ->
+            error(Reason)
+    end.
+
+-spec begin_accept(_SnapshotDir :: file:filename(), ra_snapshot:meta()) ->
+    {ok, ext_ws()}.
+begin_accept(Dir, Meta) ->
+    WS = #ws{
+        phase = machine_state,
+        started_at = erlang:monotonic_time(millisecond),
+        dir = Dir,
+        meta = Meta
+    },
+    ExtWS = #{phase => machine_state, ws => WS},
+    {ok, ExtWS}.
+
+-spec accept_chunk(binary(), ext_ws()) ->
+    {ok, ext_ws()} | {error, _Reason :: term()}.
+accept_chunk(Chunk, ExtWS = #{shard_id := ShardId}) ->
+    case gen_server:call(?via(ShardId), #accept_chunk{chunk = Chunk}, infinity) of
+        ok ->
+            {ok, ExtWS};
+        {error, Reason} ->
+            error(Reason)
+    end;
+accept_chunk(Chunk, _ExtWS = #{phase := machine_state, ws := WS0}) ->
+    %% No shard id defined yet.
+    MachineState = binary_to_term(Chunk),
+    WS = WS0#ws{state = MachineState},
+    ShardId = shard_id(MachineState),
+    case gen_server:call(?via(ShardId), #begin_accept{ws = WS}, infinity) of
+        ok ->
+            %% Transferred `ws()` to manager; no need to keep it around
+            NewExtWS = #{shard_id => ShardId},
+            {ok, NewExtWS};
+        {error, Reason} ->
+            error(Reason)
+    end.
+
+-spec complete_accept(binary(), ext_ws()) -> ok | {error, ra_snapshot:file_err()}.
+complete_accept(Chunk, _ExtWS = #{shard_id := ShardId}) ->
+    gen_server:call(?via(ShardId), #complete_accept{chunk = Chunk}, infinity).
+
+%%------------------------------------------------------------------------------
+%% `gen_server' API
+%%------------------------------------------------------------------------------
+
+init(#{db := DB, shard_id := ShardId}) ->
+    process_flag(trap_exit, true),
+    logger:set_process_metadata(#{
+        db => DB,
+        shard_id => ShardId,
+        domain => [ds, storage_layer, snapshot]
+    }),
+    clear_all_checkpoints(DB, ShardId),
+    {ok, init_state()}.
+
+terminate(_Reason, State0) ->
+    State1 = cleanup_writer(State0),
+    _ = cleanup_reader(State1),
```

Comment:
True, didn't account for the possibility of more than one reader for a shard.  :see_no_evil: 

But we should have only one writer per shard, since there's no concurrency there.